### PR TITLE
Exercise Completion Integration

### PR DESCRIPTION
**Pull request by cedricbahirwe**

### CODE REVIEW

**Comment by angiemugo** [comment]
@cedricbahirwe there's an issue I am facing, when you hit run test on an exercise, we should be able to run the tests and get some results. Right now it returns a generic error. Could you please look into this?

File: Exercism/Exercises/ExerciseEditorWindowView.swift [hunk 1]
```diff
@@ -11,26 +11,41 @@ import ExercismSwift
...
-            }.toolbar {
+            }
+            .toolbar {
                 ToolbarItem {
                     Spacer()
                 }
                 ToolbarItem(placement: .primaryAction) {
                     Button(action: {
-                        viewModel.runTest()
+                        print("Item", docs)
```
Comment (angiemugo) [inline Exercism/Exercises/ExerciseEditorWindowView.swift:47]: change this back

**Comment by angiemugo** [comment]
also, did you make some changes to `ExercismSwift`? If yes, on which branch is that work so that I can point to it?

**Comment by cedricbahirwe** [comment]
Hey @angiemugo , I did not make any changes on `ExercismSwift` that is not/ was not merged

**Comment by cedricbahirwe** [comment]
What other error are you facing?, apart from the part where I forgot to remove the print function?

**Comment by cedricbahirwe** [comment]
A question I had was: After completing an exercise, should we reload the list or do something else? (Currently doing nothing just showing an success alert)

**Comment by angiemugo** [comment]
> What other error are you facing?, apart from the part where I forgot to remove the print function?

When I run tests and one fails, I just get a generic error. I'll link it below 
Screenshot 2024-01-17 at 17 40 32

**Comment by angiemugo** [comment]
> Also, @angiemugo , @savekirk , I had a question: After completing an exercise, should we reload the list or do something else? (Currently doing nothing just showing a success alert)

I would say go back to the reloaded list of exercises

File: Exercism/Exercises/SubmitSolutionContentView.swift [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        HStack(spacing: 0) {
+            VStack(alignment: .leading) {
+                Image(systemName: "terminal")
+                    .resizable()
+                    .padding(5)
+                    .scaledToFit()
+                    .frame(width: 50, height: 40, alignment: .leading)
+                
+                VStack(alignment: .leading, spacing: 10) {
+                    Text("Publish your code and share\nyour knowledge")
```
Comment (angiemugo) [inline Exercism/Exercises/SubmitSolutionContentView.swift:43]: we have a util file where we store all the text and grab them as variables. Check `Exercism/Util/strings`
A future implementation of localisation will be required so take a look at `localizable` too

File: Exercism/Exercises/SubmitSolutionContentView.swift [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+            isSubmitting = false
+            print("Unable to complete exercise:", error)
+        }
+        
+    }
+}
+
+#Preview {
+    SubmitSolutionContentView()
+//        .preferredColorScheme(.dark)
```
Comment (angiemugo) [inline Exercism/Exercises/SubmitSolutionContentView.swift:184]: we remove this?

File: Exercism/Exercises/SubmitSolutionContentView.swift [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        do {
+            let completeSolution =  try await viewModel.completeExercise(solutionId: solution.uuid,
+                                                                         publish: shouldPublish,
+                                                                         iterationIdx: iterationIdx)
+            isSubmitting = false
+            navigationModel.goToTrack(completeSolution.track)
+            
+        } catch {
+            isSubmitting = false
+            print("Unable to complete exercise:", error)
```
Comment (angiemugo) [inline Exercism/Exercises/SubmitSolutionContentView.swift:176]: remove print statement

File: Exercism/Exercises/ExerciseEditorWindowView.swift [hunk 5]
```diff
@@ -11,20 +11,34 @@ import ExercismSwift
...
     @State var asyncModel: AsyncModel<[ExerciseFile]>
     @EnvironmentObject private var navigationModel: NavigationModel
 
     var body: some View {
         AsyncResultView(source: asyncModel) { docs in
             NavigationSplitView {
-                ExerciseRightSidebarView()
+                ExerciseRightSidebarView(
+                    canMarkAsComplete: canMarkAsComplete,
+                    onMarkAsComplete: {
```
[the remark this answers is not in the document]
Comment (cedricbahirwe) [reply Exercism/Exercises/ExerciseEditorWindowView.swift:28]: I was passing the `canMarkAsComplete` so I know whether or not to show the `Mark as complete` button?,  Do you see an alternative way to go about it?

File: Exercism/Exercises/SubmitSolutionContentView.swift (hunk 3 above)
Comment (cedricbahirwe) [reply Exercism/Exercises/SubmitSolutionContentView.swift:184]: Do you mean removing the preview or just the modifier?

File: Exercism/Exercises/ExerciseViewModel.swift [hunk 6]
```diff
@@ -108,11 +115,30 @@ final class ExerciseViewModel: ObservableObject {
...
+    
+    var sortedIterations: [Iteration] {
+        currentSolutionIterations.sorted { $0.idx > $1.idx }
+    }
+    
+    func getIterations(for solution: Solution) async {
+        do {
+            currentSolutionIterations = try await fetcher.getIterations(solution.uuid)
+        } catch {
+            print("Error getting iterations:", error)
```
[the remark this answers is not in the document]
Comment (cedricbahirwe) [reply Exercism/Exercises/ExerciseViewModel.swift:134]: what do you suggest?

File: Exercism/Exercises/ExerciseEditorWindowView.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (angiemugo) [reply Exercism/Exercises/ExerciseEditorWindowView.swift:28]: yeah, you can show this button if the closure is not nil.

File: Exercism/Exercises/ExerciseViewModel.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (angiemugo) [reply Exercism/Exercises/ExerciseViewModel.swift:134]: you can make this a throwing function and show an error to the user? or what should happen if there is an error getting an iteration?

File: Exercism/Exercises/ExerciseEditorWindowView.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (cedricbahirwe) [reply Exercism/Exercises/ExerciseEditorWindowView.swift:28]: ok

File: Exercism/Exercises/ExerciseViewModel.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (cedricbahirwe) [reply Exercism/Exercises/ExerciseViewModel.swift:134]: I think we have many [unhandled errors](https://github.com/apps-fab/exercism-app/issues/41), and we need a system to handle them (at least display them to the user), right now I'd suggest just showing the thrown error for me.

**Review by angiemugo — APPROVED** [review]

**Pull request merged**